kv/kvnemesis: TestKVNemesisMultiNode failed

kv/kvnemesis.TestKVNemesisMultiNode [failed](https://tanzanite.cluster.engflow.com/invocations/default/36760ea8-fa3d-4247-9941-ed8c012cfa48?testReportRun=5&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on release-24.1 @ [3f9e0dbd11601ab97f8a3878a0fac07ca40ba2d8](https://github.com/cockroachdb/cockroach/commits/3f9e0dbd11601ab97f8a3878a0fac07ca40ba2d8):


```
    kvnemesis_test.go:328: seed: 1568919851168614029
    kvnemesis_test.go:235: kvnemesis logging to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_25/test.outputs/kvnemesis155536639
    kvnemesis.go:178: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.ReadCommitted)
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(14559742530976203239)) // @1712552836.419034501,0 (<nil>, <nil>)
          txn.ScanForShare(ctx, tk(6535059052050322676), tk(12124352983832509520), 0) // @1712552836.419924561,0 <nil>
          txn.Put(ctx, tk(14559742530976203239), sv(2)) // @1712552836.420718201,0 <nil>
          txn.ScanForUpdate(ctx, tk(14791951014335373950), tk(17604679884215720498), 0) // @1712552836.421668872,0 <nil>
          txn.Put(ctx, tk(14559742530976203239), sv(3)) // @1712552836.422603162,0 <nil>
          txn.GetForUpdate(ctx, tk(10853573105681676505)) // @1712552836.423654292,0 (<nil>, <nil>)
          txn.CreateSavepoint(ctx, 6) // <nil>
          txn.Del(ctx, tk(14559742530976203239) /* @s4 */) // @1712552836.424492632,0 (/Table/100/"ca0e98b0ebfbc1e7", <nil>)
          txn.ReverseScanForShareGuaranteedDurability(ctx, tk(3475427587453643870), tk(11217340834543051115), 0) // @1712552836.425574962,0 <nil>
          txn.RollbackSavepoint(ctx, 6) // <nil>
          txn.GetForShareSkipLocked(ctx, tk(14559742530976203239)) // @1712552836.426497903,0 (v3, <nil>)
          txn.ScanForShareGuaranteedDurability(ctx, tk(71065404333882749), tk(1444012930575287769), 0) // @1712552836.427383603,0 <nil>
          txn.ReverseScanForUpdate(ctx, tk(1078262841857143602), tk(12817355776146110875), 0) // @1712552836.428260913,0 <nil>
          txn.ReverseScanForUpdateGuaranteedDurability(ctx, tk(5355156658035590734), tk(15827559905222623817), 0) // @1712552836.429096123,0 (/Table/100/"ca0e98b0ebfbc1e7":v3, <nil>)
          txn.GetForShare(ctx, tk(14559742530976203239)) // TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Table/100/"ca0e98b0ebfbc1e7"): "unnamed" meta={id=5296e78a key=/Table/100/"ca0e98b0ebfbc1e7" iso=ReadCommitted pri=0.11621611 epo=100 ts=1712552836.430019103,0 min=1712552835.129446061,0 seq=0} lock=true stat=PENDING rts=1712552836.430019103,0 wto=false gul=1712552835.139446061,0 isn=1
          txn.GetForShareGuaranteedDurability(ctx, tk(14559742530976203239)) // omitted
          txn.GetForUpdateGuaranteedDurability(ctx, tk(3123181576409117198)) // omitted
          txn.GetForShareSkipLocked(ctx, tk(5149875102851906418)) // omitted
          txn.GetForUpdate(ctx, tk(14559742530976203239)) // omitted
          return errors.New("rollback")
        }) // have retried transaction: unnamed (id: 5296e78a-64d8-44ca-a598-b0f4df3e9393) 101 times, most recently because of the retryable error: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Table/100/"ca0e98b0ebfbc1e7"): "unnamed" meta={id=5296e78a key=/Table/100/"ca0e98b0ebfbc1e7" iso=ReadCommitted pri=0.11621611 epo=100 ts=1712552836.430019103,0 min=1712552835.129446061,0 seq=0} lock=true stat=PENDING rts=1712552836.430019103,0 wto=false gul=1712552835.139446061,0 isn=1. Terminating retry loop and returning error due to cluster setting kv.transaction.internal.max_auto_retries (100). Rollback error: <nil>.: have retried transaction: unnamed (id: 5296e78a-64d8-44ca-a598-b0f4df3e9393) 101 times, most recently because of the retryable error: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_ASYNC_WRITE_FAILURE - missing intent on: /Table/100/"ca0e98b0ebfbc1e7"): "unnamed" meta={id=5296e78a key=/Table/100/"ca0e98b0ebfbc1e7" iso=ReadCommitted pri=0.11621611 epo=100 ts=1712552836.430019103,0 min=1712552835.129446061,0 seq=0} lock=true stat=PENDING rts=1712552836.430019103,0 wto=false gul=1712552835.139446061,0 isn=1. Terminating retry loop and returning error due to cluster setting kv.transaction.internal.max_auto_retries (100). Rollback error: <nil>.
    kvnemesis.go:213: failures(verbose): /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_25/test.outputs/kvnemesis155536639/failures
        repro steps: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_25/test.outputs/kvnemesis155536639/repro.go
        rangefeed KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_25/test.outputs/kvnemesis155536639/kvs-rangefeed.txt
        scan KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_5_of_25/test.outputs/kvnemesis155536639/kvs-scan.txt
    kvnemesis_test.go:356: [0] proposed: 2245
    kvnemesis_test.go:358: [0] reproposed unchanged: 64
    kvnemesis_test.go:360: [0] reproposed with new LAI: 0
    kvnemesis_test.go:356: [1] proposed: 163
    kvnemesis_test.go:358: [1] reproposed unchanged: 1
    kvnemesis_test.go:360: [1] reproposed with new LAI: 3
    kvnemesis_test.go:356: [2] proposed: 396
    kvnemesis_test.go:358: [2] reproposed unchanged: 1
    kvnemesis_test.go:360: [2] reproposed with new LAI: 2
    kvnemesis_test.go:356: [3] proposed: 99
    kvnemesis_test.go:358: [3] reproposed unchanged: 1
    kvnemesis_test.go:360: [3] reproposed with new LAI: 0
    kvnemesis_test.go:365: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:365
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:305
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisMultiNode
        	Messages:   	kvnemesis detected failures
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisMultiNode4067421084
--- FAIL: TestKVNemesisMultiNode (11.43s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #119681 kv/kvnemesis: TestKVNemesisMultiNode failed [A-testing C-bug C-test-failure O-robot P-3 T-kv branch-release-23.1]
- #118471 kv/kvnemesis: TestKVNemesisMultiNode failed [C-test-failure O-robot P-2 T-kv X-duplicate branch-release-23.2]
- #118017 kv/kvnemesis: TestKVNemesisMultiNode failed [A-testing C-bug C-test-failure O-robot P-2 T-kv branch-master]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37621

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_client_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner_client_test.go
@@ -15,11 +15,14 @@ import (
 	"reflect"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/isolation"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/stretchr/testify/require"
@@ -163,4 +166,32 @@ func TestTxnPipelinerCondenseLockSpans(t *testing.T) {
 		t.Fatal(err)
 	}
 	require.Zero(t, metrics.TxnsWithCondensedIntentsGauge.Value())
+}
+
+// TestTxnPipelinerLockValidationConsultsIntentHistory is a regression test for
+// #121920. The unit test verifies that an intent's intent history is consulted
+// during pipelined lock validation.
+func TestTxnPipelinerLockValidationConsultsIntentHistory(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+	db := tc.Server(0).DB()
+
+	require.NoError(t, db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		require.NoError(t, txn.SetIsoLevel(isolation.ReadCommitted))
+		key := roachpb.Key("a")
+		require.NoError(t, txn.Put(ctx, key, "val"))
+		sp, err := txn.CreateSavepoint(ctx)
+		require.NoError(t, err)
+		_, err = txn.Del(ctx, key)
+		require.NoError(t, err)
+		require.NoError(t, txn.RollbackToSavepoint(ctx, sp))
+		_, err = txn.GetForUpdate(ctx, key, kvpb.GuaranteedDurability)
+		require.NoError(t, err)
+		_, err = txn.Get(ctx, key)
+		return err
+	}))
 }

--- a/pkg/kv/kvserver/batcheval/cmd_query_intent.go
+++ b/pkg/kv/kvserver/batcheval/cmd_query_intent.go
@@ -150,7 +150,7 @@ func QueryIntent(
 			log.VEventf(ctx, 2, "found no intent")
 		}
 	} else {
-		found, err := storage.VerifyLock(
+		found, err := storage.MVCCVerifyLock(
 			ctx, reader, &args.Txn, args.Strength, args.Key, args.IgnoredSeqNums,
 		)
 		if err != nil {

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -6163,7 +6163,30 @@ func VerifyLock(
 		// nothing stopping another request from rolling back the lock even though
 		// it exists right now.
 		if enginepb.TxnSeqIsIgnored(foundLock.Txn.Sequence, ignoredSeqNums) {
-			continue // the lock is ignored, proceed to check weaker lock strengths...
+			if iterStr != lock.Intent {
+				// The lock is ignored. Proceed to check weaker lock strengths...
+				continue
+			}
+			// If the existing lock is an intent, additionally check the intent
+			// history to verify that all of the intent writes in the intent history
+			// are also rolled back. If not, an element in the intent history is
+			// providing the required protection.
+			//
+			// This is not just an optimization. It is necessary for the correctness
+			// of VerifyLock because MVCCAcquireLock will skip lock acquisition if it
+			// finds a non-rolled back intent in the intent history.
+			inHistoryNotRolledBack := false
+			for _, e := range foundLock.IntentHistory {
+				if !enginepb.TxnSeqIsIgnored(e.Sequence, ignoredSeqNums) {
+					inHistoryNotRolledBack = true
+					break
+				}
+			}
+			if !inHistoryNotRolledBack {
+				// The intent and all prior intents in the intent history are
+				// ignored, proceed to check weaker lock strengths...
+				continue
+			}
 		}
 
 		return true, nil

--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -6108,12 +6108,12 @@ func validateLockAcquisitionStrength(str lock.Strength) error {
 	return nil
 }
 
-// VerifyLock returns true if the supplied transaction holds a lock that offers
-// equal to or greater protection[1] than the supplied lock strength.
+// MVCCVerifyLock returns true if the supplied transaction holds a lock that
+// offers equal to or greater protection[1] than the supplied lock strength.
 //
 // [1] Locks that were acquired at sequence numbers that have since been ignored
 // aren't considered, as they may be rolled back in the future.
-func VerifyLock(
+func MVCCVerifyLock(
 	ctx context.Context,
 	reader Reader,
 	txn *enginepb.TxnMeta,
@@ -6173,8 +6173,8 @@ func VerifyLock(
 			// providing the required protection.
 			//
 			// This is not just an optimization. It is necessary for the correctness
-			// of VerifyLock because MVCCAcquireLock will skip lock acquisition if it
-			// finds a non-rolled back intent in the intent history.
+			// of MVCCVerifyLock because MVCCAcquireLock will skip lock acquisition if
+			// it finds a non-rolled back intent in the intent history.
 			inHistoryNotRolledBack := false
 			for _, e := range foundLock.IntentHistory {
 				if !enginepb.TxnSeqIsIgnored(e.Sequence, ignoredSeqNums) {

--- a/pkg/storage/mvcc_history_test.go
+++ b/pkg/storage/mvcc_history_test.go
@@ -1190,7 +1190,7 @@ func cmdVerifyLock(e *evalCtx) error {
 		txn := e.getTxn(optional)
 		key := e.getKey()
 		str := e.getStrength()
-		found, err := storage.VerifyLock(e.ctx, r, &txn.TxnMeta, str, key, txn.IgnoredSeqNums)
+		found, err := storage.MVCCVerifyLock(e.ctx, r, &txn.TxnMeta, str, key, txn.IgnoredSeqNums)
 		if err != nil {
 			return err
 		}

--- a/pkg/storage/testdata/mvcc_histories/verify_locks
+++ b/pkg/storage/testdata/mvcc_histories/verify_locks
@@ -256,3 +256,60 @@ found: true
 found: false
 >> at end:
 txn: "A" meta={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=20} lock=true stat=PENDING rts=10.000000000,0 wto=false gul=0,0 isn=1
+
+# Test that the intent history is consulted when verifying locks.
+run ok
+with t=A
+  txn_step seq=20
+  put k=k8 v=v8
+  txn_step seq=25
+  put k=k8 v=v8_second
+  txn_step seq=30
+  put k=k8 v=v8_third
+----
+put: lock acquisition = {k8 id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=20 Replicated Intent [{5 15}]}
+put: lock acquisition = {k8 id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=25 Replicated Intent [{5 15}]}
+put: lock acquisition = {k8 id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=30 Replicated Intent [{5 15}]}
+>> at end:
+txn: "A" meta={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=30} lock=true stat=PENDING rts=10.000000000,0 wto=false gul=0,0 isn=1
+data: "k1"/5.000000000,0 -> /BYTES/v1
+data: "k2"/5.000000000,0 -> /BYTES/v2
+data: "k3"/5.000000000,0 -> /BYTES/v3
+meta: "k4"/0,0 -> txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=10} ts=10.000000000,0 del=false klen=12 vlen=10 mergeTs=<nil> txnDidNotUpdateMeta=true
+data: "k4"/10.000000000,0 -> /BYTES/v_new
+data: "k4"/5.000000000,0 -> /BYTES/v4
+data: "k5"/5.000000000,0 -> /BYTES/v5
+data: "k6"/5.000000000,0 -> /BYTES/v6
+data: "k7"/5.000000000,0 -> /BYTES/v7
+meta: "k8"/0,0 -> txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=30} ts=10.000000000,0 del=false klen=12 vlen=13 ih={{20 /BYTES/v8}{25 /BYTES/v8_second}} mergeTs=<nil> txnDidNotUpdateMeta=false
+data: "k8"/10.000000000,0 -> /BYTES/v8_third
+
+run ok
+with t=A
+  verify_lock k=k8 str=shared
+  verify_lock k=k8 str=exclusive
+  txn_ignore_seqs seqs=(30-30)
+  verify_lock k=k8 str=shared
+  verify_lock k=k8 str=exclusive
+  txn_ignore_seqs seqs=(25-25,30-30)
+  verify_lock k=k8 str=shared
+  verify_lock k=k8 str=exclusive
+  txn_ignore_seqs seqs=(20-20,30-30)
+  verify_lock k=k8 str=shared
+  verify_lock k=k8 str=exclusive
+  txn_ignore_seqs seqs=(20-20,25-25,30-30)
+  verify_lock k=k8 str=shared
+  verify_lock k=k8 str=exclusive
+----
+found: true
+found: true
+found: true
+found: true
+found: true
+found: true
+found: true
+found: true
+found: false
+found: false
+>> at end:
+txn: "A" meta={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=10.000000000,0 min=0,0 seq=30} lock=true stat=PENDING rts=10.000000000,0 wto=false gul=0,0 isn=3

